source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -915,33 +981,68 @@ void Filter::onConnectTimeout() {
+                 read_callbacks_->connection(), data.length(), end_stream, upstream_ != nullptr,
+                 receive_before_connect_, static_cast<int>(connect_mode_));
   getStreamInfo().getDownstreamBytesMeter()->addWireBytesReceived(data.length());
+
   if (upstream_) {
     getStreamInfo().getUpstreamBytesMeter()->addWireBytesSent(data.length());
     upstream_->encodeData(data, end_stream);
     resetIdleTimer(); // TODO(ggreenway) PERF: do we need to reset timer on both send and receive?
   } else if (receive_before_connect_) {
-    ENVOY_CONN_LOG(trace, "Early data received. Length: {}", read_callbacks_->connection(),
-                   data.length());
-
     // Buffer data received before upstream connection exists.
     early_data_buffer_.move(data);
+    ASSERT(early_data_buffer_.length() > 0 || data.length() == 0);
 
-    // TCP_PROXY cannot correctly make a decision on the amount of data
-    // the preceding filters need to read before the upstream connection is established.
-    // Hence, to protect the early data buffer, TCP_PROXY read disables the downstream on
-    // receiving the first chunk of data. The filter setting the receive_before_connect state should
-    // have a limit on the amount of data it needs to read before the upstream connection is
-    // established and pause the filter chain (by returning `StopIteration`) till it has read the
-    // data it needs or a max limit has been reached.
-    read_callbacks_->connection().readDisable(true);
-
-    config_->stats().early_data_received_count_total_.inc();
+    // Track end_stream even if buffer is empty so we can propagate it to upstream later.
     if (!early_data_end_stream_) {
       early_data_end_stream_ = end_stream;
     }
+
+    // If no data was received, handle legacy mode and return early.
+    if (early_data_buffer_.length() == 0) {
+      // For legacy mode (max_buffered_bytes_ == 0), read-disable immediately even if buffer is
+      // empty. This maintains backward compatibility with the old behavior.
+      if (max_buffered_bytes_ == 0) {
+        read_callbacks_->connection().readDisable(true);
+        read_disabled_due_to_buffer_ = true;
+      }
+      return Network::FilterStatus::Continue;
+    }
+
+    // Mark that we've received initial data and trigger connection if needed.
+    if (!initial_data_received_) {
+      initial_data_received_ = true;
+      // For ON_DOWNSTREAM_DATA mode, establish the upstream connection now.
+      if (connect_mode_ == UpstreamConnectMode::ON_DOWNSTREAM_DATA) {
+        ENVOY_CONN_LOG(debug,
+                       "Initial data received, establishing upstream connection. "
+                       "early_data_buffer_.length()={}",
+                       read_callbacks_->connection(), early_data_buffer_.length());
+        // Route should already be set in onNewConnection().
+        ASSERT(route_ != nullptr);
+        establishUpstreamConnection();
+      }
+    }
```

Comment:
Seems we require the data contain the actual data, then we will trigger the connecting. But I think it will be more simpler to trigger the connecting directly here.

Then we can remove the above `if (early_data_buffer_.length() == 0) {`

Reply:
I agree. Let me change it to triggering the Connection directly here.